File: Difficulty/Median/[18] 4Sum/solution.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> res;
        int nums_size = nums.size();
        if (nums_size < 4) return res;
        sort(nums.begin(), nums.end());
        int hi = nums[nums_size - 1];
        map<int, int> num2id;
        int mapId = 0;
        for (auto& num : nums) {
            num2id[num] = mapId;
            mapId += 1;
        }
        int a = 0, b = 0, c = 0, d = 0, e = 0;
        for (int i = 0; i < nums_size - 3; i++){
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            a = nums[i];
            if (a * 4 > target) return res;
            if (a + 3 * hi < target) continue;
            for (int j = i + 1; j < nums_size - 2; j++){
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;
                b = nums[j];
                if (a + b * 3 > target) break;
                if (a + b + 2 * hi < target) continue;
                for (int k = j + 1; k < nums_size - 1; k++){
                    if (k > j + 1 && nums[k] == nums[k - 1]) continue;
                    c = nums[k];
                    if (a + b + c*2 > target) break;
                    if (a + b + c + hi < target) continue;
                    d = target - a - b - c;
                    auto search = num2id.find(d);
                    if (search != num2id.end()){
                        e = search->second;
                        if (e <= k) break;
                        else{
                            vector<int> temp{a, b, c, d};
                            res.push_back(temp);
                        }
                    }
                }
            }
        }
        return res;
    }
};

File: Difficulty/Median/[18] 4Sum/solution.py
class Solution1:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if len(nums) < 4:
            return []
        if len(nums) == 4:
            return [nums] if sum(nums) == target else []
        nums.sort()
        nums_len = len(nums)
        res = []
        for i in range(nums_len - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            for j in range(i+1, nums_len - 2):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                l, r = j + 1, nums_len - 1
                while l < r:
                    total = nums[i] + nums[j] + nums[l] + nums[r]
                    if total > target:
                        r -= 1
                    elif total < target:
                        l += 1
                    else:
                        res.append([nums[i], nums[j], nums[l], nums[r]])
                        while l < r and nums[l] == nums[l + 1]:
                            l += 1
                        while l < r and nums[r] == nums[r - 1]:
                            r -= 1
                        l += 1
                        r -= 1
        return res

# inspired by https://leetcode.com/problems/4sum/discuss/8545/Python-140ms-beats-100-and-works-for-N-sum-(Ngreater2)
class Solution2:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if len(nums) < 4:
            return []
        if len(nums) == 4:
            return [nums] if sum(nums) == target else []
        nums.sort()
        nums_len = len(nums)
        res = []
        for i in range(nums_len - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            for j in range(i+1, nums_len - 2):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                l, r = j + 1, nums_len - 1
                while l < r:
                    if (target-(nums[i]+nums[j])) > 2*nums[r] or (target-(nums[i]+nums[j])) < 2*nums[l]:
                        break
                    total = nums[i] + nums[j] + nums[l] + nums[r]
                    if total > target:
                        r -= 1
                    elif total < target:
                        l += 1
                    else:
                        res.append([nums[i], nums[j], nums[l], nums[r]])
                        while l < r and nums[l] == nums[l + 1]:
                            l += 1
                        while l < r and nums[r] == nums[r - 1]:
                            r -= 1
                        l += 1
                        r -= 1
        return res

# come from Accepted Solutions Runtime Distribution
class Solution3:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        n = len(nums)
        if n < 4:
            return []
        res = []
        nums = sorted(nums)
        hi = nums[-1]
        num2id = {x:i for i,x in enumerate(nums)}
        for i in range(n-3):
            if i > 0 and nums[i]==nums[i-1]:
                continue
            a = nums[i]
            if a*4 >target:
                return res
            if  a+3*hi<target:
                continue
            for j in range(i+1,n-2):
                if j > i+1 and nums[j]==nums[j-1]:
                    continue
                b = nums[j]
                if a+b*3 >target:
                    break
                if a+b+2*hi<target:
                    continue
                for k in range(j+1,n-1):
                    if k > j+1 and nums[k]==nums[k-1]:
                        continue
                    c = nums[k]
                    if a+b+c*2>target:
                        break
                    if a+b+c+hi<target:
                        continue
                    d = target-a-b-c
                    if d in num2id:
                        l = num2id[d]
                        if l <= k:
                            break
                        else:
                            res.append([a,b,c,d])
        return res